### PR TITLE
[maintenance] e2e test converted into Kotlin

### PR DESCRIPTION
**Pull request by basakkapusuzoglu**
BazelBspSampleRepoTest and files under base package converted from Java to Kotlin language

---

```
bazel run //e2e:BazelBspSampleRepoTest 
```

### CODE REVIEW

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            BazelBspTestBaseScenario::class.java
+        )
+        private const val SUCCESS_EXIT_CODE = 0
+        private const val FAIL_EXIT_CODE = 1
+        private const val TEST_RESOURCES_DIR = "e2e/test-resources"
+        private val WORKSPACE_DIR = System.getenv("BUILD_WORKSPACE_DIRECTORY")
+    }
+
+    init {
+        val workspacePath = getTestingRepoPath(repoName)
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:63]: u can remove it

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        exitProcess(FAIL_EXIT_CODE)
+    }
+
+    private fun executeScenarioSteps(): Boolean = scenarioSteps.stream()
+            .map { it.executeAndReturnResult() }
+            .collect(Collectors.toList())
+            .stream()
+            .allMatch { it }
+
+    protected abstract val scenarioSteps: List<BazelBspTestScenarioStep>
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:49]:
```suggestion
    protected abstract fun scenarioSteps(): List<BazelBspTestScenarioStep>
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        private const val FAIL_EXIT_CODE = 1
+        private const val TEST_RESOURCES_DIR = "e2e/test-resources"
+        private val WORKSPACE_DIR = System.getenv("BUILD_WORKSPACE_DIRECTORY")
+    }
+
+    init {
+        val workspacePath = getTestingRepoPath(repoName)
+        testClient = createClient(workspacePath)
+    }
+}
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:66]: new line

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        LOGGER.fatal("Test failed :(")
+        exitProcess(FAIL_EXIT_CODE)
+    }
+
+    private fun executeScenarioSteps(): Boolean = scenarioSteps.stream()
+            .map { it.executeAndReturnResult() }
+            .collect(Collectors.toList())
+            .stream()
+            .allMatch { it }
```
[the remark this answers is not in the document]
Comment (basakkapusuzoglu) [reply e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:47]: this is giving error

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package org.jetbrains.bsp.bazel.base
+
+import io.vavr.CheckedRunnable
+import io.vavr.control.Try
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+
+class BazelBspTestScenarioStep(private val testName: String, private val testkitCall: CheckedRunnable) {
+    fun executeAndReturnResult(): Boolean {
+        LOGGER.info("Executing \"{}\"...", testName)
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt:10]:
```suggestion
        LOGGER.info { "Executing \"$testName\"..." }
```

and u can do similar thing in each logger occurrence in edited files

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            "BspTestClient", "1.0.0", BSP_VERSION, workspacePath.toString(), capabilities
+        )
+        val client = BazelTestClient(workspacePath, initializeBuildParams)
+        LOGGER.info("Created TestClient.")
+        return client
+    }
+
+    fun executeScenario() {
+        LOGGER.info("Running scenario...")
+        LOGGER.info("Running scenario done.")
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:32]: remove it

**Review by abrams27 — CHANGES_REQUESTED** [review]
and please update `runAll.sh` file as well

File: e2e/runAll.sh [hunk 7]
```diff
@@ -14,11 +14,9 @@ runTest() {
...
 
 runTest //e2e:BazelBspSampleRepoTest
-#runTest //e2e:BazelBspActionGraphV1Test
-#runTest //e2e:BazelBspActionGraphV2Test
-#runTest //e2e:BazelBspLocalJdkTest
-#runTest //e2e:BazelBspRemoteJdkTest
-#runTest //e2e:BazelBspCppProjectTest
+runTest //e2e:BazelBspLocalJdkTest
+runTest //e2e:BazelBspRemoteJdkTest
+runTest //e2e:BazelBspCppProjectTest
```
Comment (abrams27) [inline e2e/runAll.sh:19]: this one please comment

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import io.vavr.CheckedRunnable
+import io.vavr.control.Try
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+
+class BazelBspTestScenarioStep(private val testName: String, private val testkitCall: CheckedRunnable) {
+    fun executeAndReturnResult(): Boolean {
+        LOGGER.info( "Executing \"$testName\"..." )
+        return Try.run(testkitCall)
+            .onSuccess { LOGGER.info("Step \"{}\" executed correctly!", testName) }
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt:12]: u can replace `{}` here with `$` as well

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+            .onFailure { LOGGER.error("Step \"{}\" execution failed!", testName, it) }
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt:13]: here as well

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import java.nio.file.Path
+import kotlin.system.exitProcess
+
+abstract class BazelBspTestBaseScenario(private val repoName: String?) {
+
+    protected val testClient: BazelTestClient = createClient()
+
+    private fun createClient(): BazelTestClient {
+        val workspacePath = Path.of(WORKSPACE_DIR, TEST_RESOURCES_DIR, repoName)
+        LOGGER.info("Workspace path: { \"$workspacePath\"}")
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:17]:
```suggestion
        LOGGER.info { "Workspace path: $workspacePath" }
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        LOGGER.info("Creating TestClient...")
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:18]: here as well

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        // TODO: capabilities should be configurable
+        val capabilities = BuildClientCapabilities(listOf("java", "scala", "kotlin", "cpp"))
+        val initializeBuildParams = InitializeBuildParams(
+            "BspTestClient", "1.0.0", BSP_VERSION, workspacePath.toString(), capabilities
+        )
+        val client = BazelTestClient(workspacePath, initializeBuildParams)
+        LOGGER.info("Created TestClient.")
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:25]: same

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        return client
+    }
+
+    fun executeScenario() {
+        if (executeScenarioSteps()) {
+            LOGGER.info("Test passed!")
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:31]: same

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+            exitProcess(SUCCESS_EXIT_CODE)
+        }
+        LOGGER.fatal("Test failed :(")
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:34]: same

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt (hunk 9 above)
Comment (basakkapusuzoglu) [reply e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt:13]: .onFailure { LOGGER.error("Step \"$testName\" \"$it\" execution failed!") }
is it like that?

Comment (abrams27) [reply e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt]: `.onFailure { LOGGER.error("Step \"$testName\" execution failed!", it) }`

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+import io.vavr.control.Try
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+
+class BazelBspTestScenarioStep(private val testName: String, private val testkitCall: CheckedRunnable) {
+    fun executeAndReturnResult(): Boolean {
+        LOGGER.info( "Executing \"$testName\"..." )
+        return Try.run(testkitCall)
+            .onSuccess { LOGGER.info("Step \"$testName\" executed correctly!") }
+            .onFailure  { LOGGER.error("Step \"$testName\" execution failed!", it) }
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestScenarioStep.kt:13]:
```suggestion
            .onFailure { LOGGER.error("Step \"$testName\" execution failed!", it) }
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        val client = BazelTestClient(workspacePath, initializeBuildParams)
+        LOGGER.info("Created TestClient.")
+        return client
+    }
+
+    fun executeScenario() {
+        if (executeScenarioSteps()) {
+            LOGGER.info("Test passed!")
+            exitProcess(SUCCESS_EXIT_CODE)
+        }
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/base/BazelBspTestBaseScenario.kt:33]:
```suggestion
    LOGGER.info("Running scenario...")
     val scenarioStepsExecutionResult = executeScenarioSteps()
     LOGGER.info("Running scenario done.")

     if (scenarioStepsExecutionResult) {
       LOGGER.info("Test passed!")
       exitProcess(SUCCESS_EXIT_CODE)
     }
```

**Review by abrams27 — CHANGES_REQUESTED** [review]

**Review by abrams27 — APPROVED** [review]

**Pull request merged**